### ISSUE TITLE
[meson] Eliminating the duplicated Meson trees

### ISSUE THREAD

**Issue by mcy** [lowRISC/opentitan#1162]
NOTE: the below assumes I have finished deleting Make, since the changes I wish to make cannot be expressed in Make without significant toil.

Right now, we generate one Meson tree per "platform" we support. Right now that's Verilator and FPGA, but will soon include DV, and eventually the ASIC.

I've been trying to think of a way to effectively unify these trees, for the following reasons:
- Only one ninja file to worry about, which helps make our build system less confusing to newcomers.
- Reducing the number times artifacts that do not care about the platform need to be built.
- Making it possible for Meson to know about *all* artifacts, so we don't have to figure out which Meson tree an artifact should go in. For example, we can't really have a `hw` tree, since the FPGA bitstream target needs to be able to access the bootrom target linked for FPGA mode.

All "platform-specific" behavior seems to currently be encoded in the form of constants (like clock freqs, baudrates, and similar). These constants are defined as macros with conditional compilation; this means that conditional compilation infects *everything* that pulls in those headers. My proposal is to replace header-time dependency injection with link-time dependency injection. As such, we'd replace
```c
#ifdef SIMULATION
#define CLK_HZ 0x1234
#else 
...
#endif 
```
with
```c
// platform.h
extern const size_t kClockHz;

// platform.c
#ifdef SIMULATION
const size_t kClockHz = 0x1234;
#else
...
#endif 
```

`platform.c` would have one Meson target, with corresponding defines set, for each platform we support. (e.g., `libot_platform_fpga.a`, `libot_platform_dv.a`, etc). Equivalently, we can have separate `platform_fpga.c`, `platform_dv.c` and so on that implement the declarations in `platform.h`.

Then, we only need to duplicate each OpenTitan executable for each platform we support, rather than EVERY target in the SW tree; resolving values in `platform.h` has now been delayed until link time.

I'd like to get feedback on this implementation plan; if it sounds like a good approach, I can write up a more extensive doc on how we could implement this.

**Comment by mcy** [comment]
@gkelly @asb @imphil @lenary @silvestrst

**Comment by imphil** [comment]
So that's effectively the "trick" way to insert a device tree into a binary at link time? I'm generally fine with that, we might want to to a bit further and actually doing a device-tree like struct and insert that.

@tomroberts-lowrisc did think about how we can run the same binary on different platforms, and effectively that's the first step towards that. The second step then is obviously identifying the target at runtime, which requires a bit of HW support.

**Comment by mcy** [comment]
Yeah, the exact contents of `platform.h` are a bit wobbly in my head. I think that's the less important detail, honestly. I'm mostly interested in the "dependency injection" mechanism. It can even be at runtime in the future, but I mostly just want to kill the split Meson trees as soon as possible.

**Comment by lenary** [comment]
This sounds reasonable, and should mostly accomplish the primary goal.

I suppose where it falls down is if we have compiler flags that we want the entire tree compiled with. For instance, we usually want the whole tree compiled with `-Oz` for normal software, but sometimes (for example benchmarking), we might want the whole tree compiled with `-O3` (and a bunch of other optimisation flags). Note these flags cannot just be link-time, they have to be provided from the time you turn a .c into a .o file. 

We don't really have a solution for this now, and it's not clear that you can do this except with two different trees of libraries, so it might just be something we delay for the moment, which is fine.

**Comment by asb** [comment]
Moving to a single Meson tree is definitely a positive step. The advantage of the current define-based approach for these constants is that it's fairly intuitive what will be evaluated at compile time etc. For a non-LTO build, this change may cause some minor code size regressions. But if that becomes a problem, we can always address it by enabling LTO.

**Comment by mcy** [comment]
I wrote a doc last year about this but the holidays got in the way. Here it is: https://docs.google.com/document/d/1dPn0RDTG2cx1eN2U8QkkY299tLDn-DUl_wVFyWDp5Eg

There's a few places that are incomplete, and I think comments got pruned when I copied it over to the OT org, but it expresses the key sentiments.

@gkelly @lenary @asb @imphil @sriyerg

**Comment by mcy** [comment]
This was fixed by #1433 but it never got closed.
